app/services/shop_products_service.rb

```diff
@@ -0,0 +1,76 @@
+# Returns a (paginatable) AR object for the products or variants in stock for a given shop and OC.
+# The stock-checking includes on_demand and stock level overrides from variant_overrides.
+
+class ShopProductsService
+  def initialize(distributor, order_cycle)
+    @distributor = distributor
+    @order_cycle = order_cycle
+  end
+
+  def products_relation
+    Spree::Product.where(id: stocked_products)
+  end
+
+  def variants_relation
+    @order_cycle.
+      variants_distributed_by(@distributor).
+      merge(stocked_variants_and_overrides)
+  end
+
+  private
+
+  def stocked_products
+    @order_cycle.
+      variants_distributed_by(@distributor).
+      merge(stocked_variants_and_overrides).
+      select("DISTINCT spree_variants.product_id")
+  end
+
+  def stocked_variants_and_overrides
+    Spree::Variant.
+      joins("LEFT OUTER JOIN variant_overrides ON variant_overrides.variant_id = spree_variants.id
+            AND variant_overrides.hub_id = #{@distributor.id}").
+      joins(:stock_items).
+      where(query_stock_with_overrides)
+  end
+
+  def query_stock_with_overrides
```

Comment:
I guess this eager loads data that would have been loaded with N+1 by `ScopeVariantToHub` otherwise?

Reply:
Yes, it includes the post-scoping stock check in the main query so that it works when paginating results. Otherwise if we fetch 10 products first, then scope, then check stock with overrides and remove the out-of-stock products, the paginated results might return 4 or 7, or whatever, instead of 10. Does that make sense?